src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+                }
+
+                return ValueTask.FromResult<RegisterExtensionResponse>(new(
+                    assemblyHandlers.WorkspaceMessageHandlers.Keys.ToImmutableArray(),
+                    assemblyHandlers.DocumentMessageHandlers.Keys.ToImmutableArray()));
+            }
+        }
+
+        return extension.LoadAssemblyAsync(assemblyFileName);
+    }
+
+    public ValueTask<string> HandleExtensionWorkspaceMessageAsync(Solution solution, string messageName, string jsonMessage, CancellationToken cancellationToken)
+        => HandleExtensionMessageAsync(solution, messageName, jsonMessage, _workspaceHandlers, cancellationToken);
+
+    public ValueTask<string> HandleExtensionDocumentMessageAsync(Document document, string messageName, string jsonMessage, CancellationToken cancellationToken)
+        => HandleExtensionMessageAsync(document, messageName, jsonMessage, _documentHandlers, cancellationToken);
+
+    public ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        Extension? extension = null;
+        lock (_lockObject)
+        {
+            if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                if (extension.RemoveAssemblyHandlers(assemblyFileName, out var assemblyHandlers))
+                {
+                    if (assemblyHandlers is not null)
+                    {
+                        foreach (var workspaceHandler in assemblyHandlers.WorkspaceMessageHandlers.Keys)
+                        {
+                            _workspaceHandlers.Remove(workspaceHandler);
+                        }
+
+                        foreach (var documentHandler in assemblyHandlers.DocumentMessageHandlers.Keys)
+                        {
+                            _documentHandlers.Remove(documentHandler);
+                        }
+                    }
+                }
+
+                if (extension.AssemblyHandlersCount > 0)
+                {
+                    return ValueTask.CompletedTask;
+                }
+
+                _extensions.Remove(assemblyFolderPath);
+            }
+        }
+
+        extension?.AnalyzerAssemblyLoader.Dispose();
+
+        return ValueTask.CompletedTask;
+    }
+
+    public void Reset()
+    {
+        List<Extension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _workspaceHandlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AnalyzerAssemblyLoader.Dispose();
+        }
+    }
+
+    public void Dispose()
```

Comment:
as per prior convo, consider not being Disposable.  i don't see it providing value, and it adds more complexity/confusion.  

note: i would expect that at the very least, once disposed, this type would not be usable after that point.